tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php

```diff
@@ -0,0 +1,663 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Reflection;
+
+use JetBrains\PHPStormStub\PhpStormStubsMap;
+use PhpParser\Node;
+use PhpParser\Node\Name;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitorAbstract;
+use PHPStan\Parser\Parser;
+use PHPStan\Php8StubsMap;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Testing\PHPStanTestCase;
+use PHPStan\Type\VerbosityLevel;
+use ReflectionClass;
+use Symfony\Component\Finder\Finder;
+use Throwable;
+use function array_keys;
+use function array_merge;
+use function count;
+use function dirname;
+use function explode;
+use function file_get_contents;
+use function file_put_contents;
+use function floor;
+use function get_class;
+use function get_declared_classes;
+use function get_defined_functions;
+use function getenv;
+use function implode;
+use function mkdir;
+use function sort;
+use function strpos;
+use function trim;
+use const PHP_INT_MAX;
+use const PHP_VERSION_ID;
+
+class ReflectionProviderGoldenTest extends PHPStanTestCase
+{
+
+	/** @return iterable<string, array<string>> */
+	public static function data(): iterable
+	{
+		$inputFile = self::getTestInputFile();
+		$contents = file_get_contents($inputFile);
+
+		if ($contents === false) {
+			self::fail('Input file \'' . $inputFile . '\' is missing.');
+		}
+
+		$parts = explode('-----', $contents);
+
+		for ($i = 1; $i + 1 < count($parts); $i += 2) {
+			$input = trim($parts[$i]);
+			$output = trim($parts[$i + 1]);
+
+			yield $input => [
+				$input,
+				$output,
+			];
+		}
+	}
+
+	/** @dataProvider data */
+	public function test(string $input, string $expectedOutput): void
+	{
+		$output = self::generateSymbolDescription($input);
+		$output = trim($output);
+		$this->assertSame($expectedOutput, $output);
+	}
+
+	private static function generateSymbolDescription(string $symbol): string
+	{
+		[$type, $name] = explode(' ', $symbol);
+
+		try {
+			switch ($type) {
+				case 'FUNCTION':
+					return self::generateFunctionDescription($name);
+				case 'CLASS':
+					return self::generateClassDescription($name);
+				case 'METHOD':
+					return self::generateClassMethodDescription($name);
+				case 'PROPERTY':
+					return self::generateClassPropertyDescription($name);
+				default:
+					self::fail('Unknown symbol type ' . $type);
+			}
+		} catch (Throwable $e) {
+			// phpcs:ignore SlevomatCodingStandard.Namespaces.ReferenceUsedNamesOnly
+			if ($e instanceof \PHPUnit\Exception) {
+				throw $e;
+			}
+
+			$result = "Generating symbol description failed:\n"
```

Comment:
Ok, less work for me. It will make it a bit more involved do debug these things, but it's not the main purpose of the test anyway.

These symbols fail on PHP 7.4 (7.3 also fails, though I haven't checked which symbols - presumably it's the same ones):

```
CLASS Random\IntervalBoundary
CLASS Relay\KeyType
METHOD Random\IntervalBoundary::cases
METHOD Random\Randomizer::getFloat
PROPERTY Random\IntervalBoundary::$name
```

It seems like a better-reflection bug (or maybe it's a feature). There are probably going to be bugs like that from time to time, so IMO it's better not to break the whole test because of it. There are two distinct error messages:

```
-----
CLASS Random\IntervalBoundary
-----
Generating symbol description failed:
PHPStan\BetterReflection\Reflector\Exception\IdentifierNotFound: PHPStan\BetterReflection\Reflection\ReflectionClass "UnitEnum" could not be found in the located source in /app/vendor/ondrejmirtes/better-reflection/src/Reflector/Exception/IdentifierNotFound.php:31
Stack trace:
#0 /app/vendor/ondrejmirtes/better-reflection/src/Reflector/DefaultReflector.php(40): PHPStan\BetterReflection\Reflector\Exception\IdentifierNotFound::fromIdentifier(Object(PHPStan\BetterReflection\Identifier\Identifier))
#1 /app/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php(1291): PHPStan\BetterReflection\Reflector\DefaultReflector->reflectClass('UnitEnum')
#2 /app/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php(1654): PHPStan\BetterReflection\Reflection\ReflectionClass->addEnumInterfaces(Array)
#3 /app/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php(1477): PHPStan\BetterReflection\Reflection\ReflectionClass->getCurrentClassImplementedInterfacesIndexedByName()
#4 /app/vendor/ondrejmirtes/better-reflection/src/Reflection/Adapter/ReflectionClass.php(392): PHPStan\BetterReflection\Reflection\ReflectionClass->getInterfaces()
#5 /app/src/Reflection/ClassReflection.php(879): PHPStan\BetterReflection\Reflection\Adapter\ReflectionClass->getInterfaces()
#6 /app/tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php(245): PHPStan\Reflection\ClassReflection->getImmediateInterfaces()
#7 /app/tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php(79): PHPStan\Reflection\ReflectionProviderGoldenTest::generateClassDescription('Random\\Interval...')
#8 /app/tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php(112): PHPStan\Reflection\ReflectionProviderGoldenTest::generateSymbolDescription('CLASS Random\\In...')
#9 /app/tests/generate-reflection-test.php(10): PHPStan\Reflection\ReflectionProviderGoldenTest::dumpOutput()
#10 {main}
-----
METHOD Random\Randomizer::getFloat
-----
Generating symbol description failed:
LogicException: This enum case does not have a value in /app/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionEnumCase.php:111
Stack trace:
#0 /app/vendor/ondrejmirtes/better-reflection/src/Reflection/Adapter/ReflectionClassConstant.php(100): PHPStan\BetterReflection\Reflection\ReflectionEnumCase->getValueExpression()
#1 /app/src/Reflection/InitializerExprTypeResolver.php(1891): PHPStan\BetterReflection\Reflection\Adapter\ReflectionClassConstant->getValueExpression()
#2 /app/src/Reflection/InitializerExprTypeResolver.php(1946): PHPStan\Reflection\InitializerExprTypeResolver->getClassConstFetchTypeByReflection(Object(PhpParser\Node\Name\FullyQualified), 'ClosedOpen', Object(PHPStan\Reflection\ClassReflection), Object(Closure))
#3 /app/src/Reflection/InitializerExprTypeResolver.php(174): PHPStan\Reflection\InitializerExprTypeResolver->getClassConstFetchType(Object(PhpParser\Node\Name\FullyQualified), 'ClosedOpen', 'Random\\Randomiz...', Object(Closure))
#4 /app/src/Reflection/Php/PhpParameterReflection.php(114): PHPStan\Reflection\InitializerExprTypeResolver->getType(Object(PhpParser\Node\Expr\ClassConstFetch), Object(PHPStan\Reflection\InitializerExprContext))
#5 /app/tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php(374): PHPStan\Reflection\Php\PhpParameterReflection->getDefaultValue()
#6 /app/tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php(307): PHPStan\Reflection\ReflectionProviderGoldenTest::generateVariantsDescription('getFloat', Array)
#7 /app/tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php(81): PHPStan\Reflection\ReflectionProviderGoldenTest::generateClassMethodDescription('Random\\Randomiz...')
#8 /app/tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php(112): PHPStan\Reflection\ReflectionProviderGoldenTest::generateSymbolDescription('METHOD Random\\R...')
#9 /app/tests/generate-reflection-test.php(10): PHPStan\Reflection\ReflectionProviderGoldenTest::dumpOutput()
#10 {main}
```

Reply:
It's because it's unrealistic for these classes to exist on these versions. PHPStan already has that information, you might filter that out when generating the symbol list.

`PhpStormStubsSourceStubber` has `isPresentClass` method that respects the `$phpVersion` passed into the constructor.

If you want to utilize such PHPStan/BetterReflection internals in the generator script, you can do that by creating `new ContainerFactory` and then getting the container by calling `->create`.

Reply:
> It's because it's unrealistic for these classes to exist on these versions.

Yeah, I expected as much. But I'd prefer to keep the symbols "from the future" (or past) in there (even if they generate an exception as is the case here). That will prevent us from accidentally introducing these symbols into older PHP versions. E.g. I currently have a dump for PHP 7.4 generated and this is in there:

```php
FUNCTION str_starts_with
-----
Has side-effects: Maybe
NOT BUILTIN
Variants: 1
    /**
     * @param string|null $haystack
     * @param string|null $needle
     * @return bool
     */
    function str_starts_with(string|null $haystack, string|null $needle): bool
```
I haven't checked how it got there (maybe a polyfill - since it says NOT BUILTIN?), but ideally it shouldn't be there.